// @nrwl/schematics/src/lib-versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.angularCliVersion = '7.0.1';
exports.angularVersion = '^7.0.0';
exports.angularDevkitVersion = '~0.10.0';
exports.angularJsVersion = '1.6.6';
exports.ngrxVersion = '6.1.0';
exports.ngrxStoreFreezeVersion = '0.2.4';
exports.nxVersion = '7.0.2';
exports.schematicsVersion = '7.0.2';
exports.latestMigration = '20180507-create-nx-json';
exports.prettierVersion = '1.13.7';
exports.typescriptVersion = '~3.1.3';
exports.rxjsVersion = '~6.3.3';
exports.jestVersion = '^23.0.0';
exports.jestPresetAngularVersion = '6.0.1';
exports.jasmineMarblesVersion = '0.4.0';
exports.cypressVersion = '^3.1.0';
exports.expressVersion = '4.16.3';
exports.expressTypingsVersion = '4.16.0';
exports.libVersions = {
    angularVersion: exports.angularVersion,
    angularDevkitVersion: exports.angularDevkitVersion,
    angularCliVersion: exports.angularCliVersion,
    angularJsVersion: exports.angularJsVersion,
    ngrxVersion: exports.ngrxVersion,
    ngrxStoreFreezeVersion: exports.ngrxStoreFreezeVersion,
    nxVersion: exports.nxVersion,
    schematicsVersion: exports.schematicsVersion,
    prettierVersion: exports.prettierVersion,
    latestMigration: exports.latestMigration,
    typescriptVersion: exports.typescriptVersion,
    rxjsVersion: exports.rxjsVersion,
    jasmineMarblesVersion: exports.jasmineMarblesVersion
};
